commerce/domain/shipping/cart/calculator/classes/ShippingCartCalculatorSample.cls

```diff
@@ -51,15 +55,170 @@ public class ShippingCartCalculatorSample extends CartExtension.ShippingCartCalc
-          CartExtension.CartDeliveryGroupMethod cartDeliveryGroupMethod02 = new CartExtension.CartDeliveryGroupMethod('Next Day Air', 15.99, shippingProduct);
-          cartDeliveryGroupMethod02.setCarrier('UPS');
-          cartDeliveryGroupMethod02.setClassOfService('Next Day Air');
-          cartDeliveryGroupMethods.add(cartDeliveryGroupMethod01);
-          cartDeliveryGroupMethods.add(cartDeliveryGroupMethod02);
+            if(useHTTPService){
+                // Get shipping options, including aspects like rates and carriers, from the external service.
+                ShippingOptionsAndRatesFromExternalService[] shippingOptionsAndRatesFromExternalService = getShippingOptionsAndRatesFromExternalService(
+                  numberOfUniqueItems, cartValidationOutputList
+                  );
+
+                  // Create a CartDeliveryGroupMethod record for every shipping option returned from the external
+                  // service and every Order Delivery Method that matches
+                 if(shippingOptionsAndRatesFromExternalService != null){
+                  populateCartDeliveryGroupMethodWithShippingOptions(
+                    shippingOptionsAndRatesFromExternalService,
+                    cartDeliveryGroupMethods,shippingProduct,
+                    cartValidationOutputList
+                  );
+                }
+            } else {
+              CartExtension.CartDeliveryGroupMethod cartDeliveryGroupMethod01 = new CartExtension.CartDeliveryGroupMethod('Ground Shipping', 10.99, shippingProduct);
+              cartDeliveryGroupMethod01.setCarrier('USPS');
+              cartDeliveryGroupMethod01.setClassOfService('Ground Shipping');
+              CartExtension.CartDeliveryGroupMethod cartDeliveryGroupMethod02 = new CartExtension.CartDeliveryGroupMethod('Next Day Air', 15.99, shippingProduct);
+              cartDeliveryGroupMethod02.setCarrier('UPS');
+              cartDeliveryGroupMethod02.setClassOfService('Next Day Air');
+              cartDeliveryGroupMethods.add(cartDeliveryGroupMethod01);
+              cartDeliveryGroupMethods.add(cartDeliveryGroupMethod02);
+            }
+         }
       }
+  }
+
+   private static String generateRandomString(Integer len) {
+        final String chars = 'ABCDEFGHIJKLMNOPQRSTUVWXYZ0123456789abcdefghijklmnopqrstuvwxyz';
+        String randStr = '';
+        while (randStr.length() < len) {
+        Integer idx = Math.mod(Math.abs(Crypto.getRandomInteger()), chars.length());
+        randStr += chars.substring(idx, idx+1);
+        }
+        return randStr;
+    }
+
+  // Note: This sample method currently only takes in numberOfUniqueItems as an input parameter. For
+  // real-world scenarios, expand the parameter list.
+  private ShippingOptionsAndRatesFromExternalService[] getShippingOptionsAndRatesFromExternalService(
+    Integer numberOfUniqueItems, CartExtension.CartValidationOutputList cartValidationOutputCollection) {
```

Comment:
What is numberOfUniqueItems for?

Reply:
This is to indicate that multiple parameters can be sent as an input to the externalService